### ISSUE TITLE
Encoder might skip intermediate resources when fieldset filters are applied

### ISSUE THREAD

**Issue by hglattergotz** [neomerx/json-api#98]
At the application level a developer is presented with the request that might look like this:

``` http
GET /sites/1?include=posts.comments,comments.author&fields[comments]=body HTTP/1.1
Accept: application/vnd.api+json
```

From the test example [EncodeIncludedObjectsTest](https://github.com/neomerx/json-api/blob/master/tests/Encoder/EncodeIncludedObjectsTest.php#L159) we can see that the `include` and `fields` parameters need to be combined (in this example `comments`):

``` php
    public function testEncodeWithRecursiveIncludedObjects()
    {
        $this->author->{Author::LINK_COMMENTS} = $this->comments;

        $actual = Encoder::instance([
            Author::class  => AuthorSchema::class,
            Comment::class => CommentSchema::class,
            Post::class    => PostSchema::class,
            Site::class    => SiteSchema::class,
        ], $this->encoderOptions)->encodeData($this->site, new EncodingParameters(
            // include only this relation (according to the spec intermediate will be included as well)
            [Site::LINK_POSTS . '.' . Post::LINK_COMMENTS],
            // include only these attributes and links
            [
                'comments' => [Comment::ATTRIBUTE_BODY, Comment::LINK_AUTHOR],
                'posts'    => [Post::LINK_COMMENTS],
                'sites'    => [Site::LINK_POSTS],
            ]
        ));
        :
        :
    }
```

Given that the url will contain the includes and fields in separate parameters, it is up to the application level to correctly parse `posts.comments,comments.author` and figure out to create the following for the `$includePaths`:

``` php
    [Site::LINK_POSTS . '.' . Post::LINK_COMMENTS], // posts.comments
```

and

``` php
    'comments' => [Comment::ATTRIBUTE_BODY, Comment::LINK_AUTHOR /*author*/],
```

for `$fieldSets`.

Would it not be better to put this logic in the library?

**Comment by neomerx** [comment]
Do you mean to have separate parameters for them? Can you post a code piece how its usage could look like?

**Comment by hglattergotz** [comment]
Looking through the code base and tests a bit more I found `Neomerx\JsonApi\Parameters\ParametersParser`.

Using the example in this issue

``` http
GET /sites/1?include=posts.comments,comments.author&fields[comments]=body HTTP/1.1
Accept: application/vnd.api+json
```

Parsing this with an instance of `ParametersParser` I believe you would end up with a `Parameters` instance with the following private member values:

``` php
$this->includePaths = ['posts.comments','comments.author'];
$this->fieldSets = ['comments' => ['body']]
```

You would NOT get

``` php
$this->includePaths = ['posts.comments'];
$this->fieldSets = ['comments' => ['body', 'author']]
```

What I was asking originally is if there is code in this library that will build the EncodingParameters object for me from the query string. The answer to that is YES: `ParametersParser`
But, I think that ParametersParser does not output an instance of EncodingParameters/Parameters that would be needed to make the test pass because the parser does not add links to the field sets (it will not create `['comments' => ['body', 'author']`, but instead will create `['comments' => ['body']` which will cause the Encode not to include the Author).

**Comment by lindyhopchris** [comment]
I think @hglattergotz interpretation is correct, however the include links shouldn't be added to the fields asked for by the client. (It's unsafe to start adding to what the client has actually asked for.) Instead the encoder needs to process what fields to encode independently from what included resources need to be encoded. 

The definition of fields is:

> A resource object's attributes and its relationships are collectively called its "fields".

So the example request has asked for comments to only have the `body` field included, but has not asked for the `author` relationship field to be included. So

``` php
$this->includePaths = ['posts.comments','comments.author'];
$this->fieldSets = ['comments' => ['body']]
```

is correct based on the request received. The request has not asked for the `comments.author` relationship field to be included. (Remember the relationship field is the identifier contained within the comments resource object - not the related resource.)

What should now happen is that because `comments.author` has been requested as an include, the encoder needs to include that related resource in the returned response even though any encoded comments resource will not have their author relationship field encoded.

I.e. the logic as to whether a related resource is included in the encoded response is independent from whether the relationship field is being encoded.

So your comment would look like this:

``` json
{
    "type": "comments",
    "id": "123",
    "attributes": {
        "body": "The comment body"
    }
}
```

But the included resources should included the comment's author resource.

It does not matter that the comments.author relationship field is not included in the returned compound document because:

> The only exception to the full linkage requirement is when relationship fields that would otherwise contain linkage data are excluded via sparse fieldsets.

**Comment by hglattergotz** [comment]
@lindyhopchris It did not even occur to me that relationship names could be included in the sparse field set parameter or the `fields[type]` url parameter. I though this was a neomerx/json-api'ism.

> A resource object's attributes and its relationships are collectively called its "fields".

The above does not explicitly say either way if this applies to the `fields[type]` url parameter (IMO). But various discussions on this subject seem to indicate that relationships can be both in `include` and `fields[type]`.

Your summary seems on point (@lindyhopchris).

So it seems like because of my lack of understanding I inadvertently found an issue. ;-)

**Comment by lindyhopchris** [comment]
@hglattergotz 

The quote definitely does refer to the `fields` parameter. The spec says:

> A client MAY request that an endpoint return only specific fields in the response on a per-type basis by including a fields[TYPE] parameter.
> http://jsonapi.org/format/#fetching-sparse-fieldsets

The word `fields` in that quote is hyper-linked to the definition of `fields`, which is:

> A resource object's attributes and its relationships are collectively called its "fields".
> http://jsonapi.org/format/#document-resource-object-fields

I'd agree, you seem to have found an issue!

**Comment by neomerx** [comment]
Can you please explain the issue in plain English?

**Comment by lindyhopchris** [comment]
Say there's this set of data:

``` http
GET /posts/1?include=comments.author

{
    "data": {
        "type": "posts",
        "id": "1",
        "attributes": {
            "title": "My post",
            "body": "Post body...",
            "date": "2015-01-01"
        },
        "relationships": {
            "comments": {
                "data": [
                    {
                        "type": "comments",
                        "id": "123"
                    }
                ]
            }
        }
    },
    "included": [
        {
            "type": "comments",
            "id": "123",
            "attributes": {
                "body": "My comment...",
                "date": "2015-01-01",
                "likes": 5
            },
            "relationships": {
                "author": {
                    "data": {
                        "type": "people",
                        "id": "345"
                    }
                }
            }
        },
        {
            "type": "people",
            "id": "345",
            "attributes": {
                "name": "Foo",
                "surname": "Bar"
            }
        }
    ]
}
```

If only the comment's `body` field is requested, then the result should be:

``` http
GET /posts/1?include=comments.author&fields[comments]=body

{
    "data": {
        "type": "posts",
        "id": "1",
        "attributes": {
            "title": "My post",
            "body": "Post body...",
            "date": "2015-01-01"
        },
        "relationships": {
            "comments": {
                "data": [
                    {
                        "type": "comments",
                        "id": "123"
                    }
                ]
            }
        }
    },
    "included": [
        {
            "type": "comments",
            "id": "123",
            "attributes": {
                "body": "My comment..."
            }
        },
        {
            "type": "people",
            "id": "345",
            "attributes": {
                "name": "Foo",
                "surname": "Bar"
            }
        }
    ]
}
```

I.e. author 345 should still be in the `included` member even though the `comments.relationships.author` field is not included in the comment.

@hglattergotz correct me if I've got it wrong, but I believe you're saying that the encoder would actually not include the `author` resource in the `included` member because the `comments.author` field has not been requested?

**Comment by neomerx** [comment]
@lindyhopchris thanks for explanation

`feature/issue98` branch created

**Comment by hglattergotz** [comment]
@lindyhopchris yes, that is correct

**Comment by hglattergotz** [comment]
@neomerx Would it help if I created a few tests to demonstrate what I think is not working correctly?

If so, I think the most useful test would be something like this
- Create a mock request object with headers and query string
- Parse the request into a `Neomerx\JsonApi\Parameters\Parameters` instance using the `ParameterParser`
- Use `Encoder::encodeData()` with the above Parameters to create the json output
- Compare the output json to expected json

This would be an expanded version of `tests/Sample`.

Should I just let you know when I have it pushed to remote on my fork (so you can see if this is how you would want it done) or would you want me to make a pull request against feature/issue98?

**Comment by neomerx** [comment]
@hglattergotz Thanks. Having a test in `feature/issue98` would be great. I think [one of these](https://github.com/neomerx/json-api/blob/master/tests/Encoder/EncodeSparseAndFieldSetsTest.php) could be a good starting point.

**Comment by hglattergotz** [comment]
@neomerx I took a stab at it and pushed to my [fork](https://github.com/hglattergotz/json-api/blob/429a44ac05e331e5b5b1e6dd8cd471df92971a13/tests/Integration/DefaultsTest.php). Let me know if that is ok like that and if you want me to open a pull request.

**Comment by neomerx** [comment]
I think it should be added to the file I've mentioned. Using of already existing models and etc will reduce the test a few times.

**Comment by hglattergotz** [comment]
The "problem" with the existing modes/schemas is, they have default links/relationships. For example, the SiteSchema by default will include `posts`, `posts.author` and `posts.comments` which changes the behavior of the EncodingParameters (I think).
In other words, I have no way to control whether posts gets included or not. It always will be.

The problem I am trying to highlight is that if the `/sites/` endpoint would only include site resources by default, and I would ask for `/sites/1?include=posts&fields[posts]=title` the whole thing breaks.

**Comment by neomerx** [comment]
I've added [this line](https://github.com/neomerx/json-api/blob/feature/issue98/tests/Encoder/EncodeSparseAndFieldSetsTest.php#L283) to the test. If uncommented the test will pass. So it's about attributes filtering logic which I believe is located [here](https://github.com/neomerx/json-api/blob/feature/issue98/src/Schema/ResourceObject.php#L158-L169). (apply filter)

**Comment by neomerx** [comment]
Actual logic which fieldset filter to be applied to type `X` is [here](https://github.com/neomerx/json-api/blob/feature/issue98/src/Parameters/EncodingParameters.php#L66-L75) (choose filter)

**Comment by neomerx** [comment]
I've updated the filter logic.
**Previously**
- if no fieldeset filter is given then all fields were in the result.
- if at a filter is given for at least one type then by default all other types had filter that removed everything (no fields in output)

**Now**
- types that do not have filter set won't be filtered (all fields will be in output)

I looks more reasonable behavior and seems to solve the problem.

Sorry for a small delay. I've had problems with my computer.

If no objections from your side I'll squash and merge into `develop` and then to `master`.

**Comment by neomerx** [comment]
I especially like how 'big' the fix was. An `[]` was replaced with `null` [in this line](https://github.com/neomerx/json-api/blob/feature/issue98/src/Parameters/EncodingParameters.php#L73) :laughing:

**Comment by hglattergotz** [comment]
Thanks for solving this. I did not have any time to look into this and assist.
What you have explained seems correct and is what I would expect to see.

Thanks!
Glad it was "easy"

**Comment by neomerx** [comment]
merged to `develop`

**Comment by hglattergotz** [comment]
Unfortunately I found a case where this is still not fixed:

If you add a fields filter for the `sites` resource the include is suppressed (but should not be).

If you add `'sites' => [Site::ATTRIBUTE_NAME],` to line 302 in [this file](https://github.com/neomerx/json-api/blob/master/tests/Encoder/EncodeSparseAndFieldSetsTest.php#L302) the include of `posts` is ignored.

**Comment by neomerx** [comment]
I don't agree. If you ask encoder to keep in 'sites' only attributes but not relationships (don't forget they are a fieldset) you will get exactly that - a `Site` with attributes only.

``` json
{
    "data":{
        "type":"sites",
        "id":"2",
        "attributes":{
            "name":"site name"
        },
        "links":{
            "self":"http://example.com/sites/2"
        }
    }
}
```

**Comment by hglattergotz** [comment]
hmmm, maybe I am not understanding this correctly then. But I would expect

``` html
GET /sites/2?include=posts&fields[sites]=name&fields[posts]=title,body
```

to give me the output from test case `testIncludeAndSparseFieldSets`, no? Because of `include=posts`

**Comment by hglattergotz** [comment]
This was the point that @lindyhopchris also argued earlier, I think.

**Comment by neomerx** [comment]
Such a request has a logical error. It says
- `fields[sites]=name` - I don't want any relationships, attributes only
- `include=posts` - I want relationships

**Comment by neomerx** [comment]
If you want to have a relationship in `main data` you keep in in fieldset filter. For example

``` php
    'sites' => [Site::ATTRIBUTE_NAME, Site::LINK_POSTS],
```

**Additionally** if you want resources from this relationship to be included you ask it to be included

``` php
    Site::LINK_POSTS
```

**Comment by hglattergotz** [comment]
Maybe the `relationships` part in the `sites` would be excluded, but the `posts` should still be in the `included` section.

So, what you are saying is that if you omit a field set on sites it takes the defaults (all), which would technically be `fields[sites]=name,posts`?
And as soon as you explicitly add `fields[sites]` you have to include relationships explicitly.

**Comment by neomerx** [comment]
It's not possible. It will break ["full linkage"](http://jsonapi.org/format/#document-compound-documents) spec concept.

**Comment by neomerx** [comment]
and there "The only exception to the full linkage requirement is when relationship fields that would otherwise contain linkage data are excluded via sparse fieldsets." lol

**Comment by neomerx** [comment]
but anyway current behavior is within the spec

**Comment by hglattergotz** [comment]
I am not sure I agree, but I do not want to suck up any more of your time on this. I know you are probably super busy and I appreciate all your work here.
I will try to research this and find some definitive answers.

I really wish they had a language agnostic test suite with sample data in the json-api project so we don't have to figure this out.

Thanks

**Comment by neomerx** [comment]
Yes, if you omit filter it will mean _all fields_

**Comment by hglattergotz** [comment]
@neomerx, @lindyhopchris I checked on the discussion forum for json-api on this matter and am now certain that this is in fact still an issue.
Please check out the following detailed [response](http://discuss.jsonapi.org/t/clarification-on-the-sparse-fieldset-response/221/11) to my [question](http://discuss.jsonapi.org/t/clarification-on-the-sparse-fieldset-response/221/7)

**Comment by hglattergotz** [comment]
@neomerx :point_up:

**Comment by neomerx** [comment]
I'm working on next version of Limoncello (and json-api) so I revisit this issue within a couple of weeks.

**Comment by neomerx** [comment]
@hglattergotz I've re-read the comments and as I can see you were discussed if 'full-linkage' rule could be broken. You concluded it could be broken in filtered field-sets but it would be a bad practice.

>  but it's probably a bad practice, since it makes it harder for the client to connect the included resources to the primary data in some cases.

Neither the spec, nor the answers do not forbid providing 'full-linkage' in such cases. Moreover it could be considered as good practice as it makes easier to connect resources. Therefore I don't see any issue on this with the current implementation.

**Comment by lindyhopchris** [comment]
@neomerx 

I disagree. The spec is clear that the full-linkage does not apply in this scenario:

> The only exception to the full linkage requirement is when relationship fields that would otherwise contain linkage data are excluded via sparse fieldsets.

To include full linkage when the client has not asked for it is bad API design: the client has made a specific request, the response should adhere to that request rather than making executive decisions as to what should be included regardless of what the client has specified.

**Comment by neomerx** [comment]
As I understand this 'exception' it is a right not an obligation to break 'full linkage' rule

**Comment by neomerx** [comment]
To make it clear 'full linkage' doesn't mean all intermediate resources are included. Only their identities (id, type but not fields) present in output.

**Comment by lindyhopchris** [comment]
> are excluded via sparse fieldsets

i.e. relationship fields (type, id) are not present because the client has asked for them _to be excluded_. Therefore it is an obligation based on what the client has asked for.

I'm worried this is deviating from the spec, particularly after reading the response to @hglattergotz question on the forum.

**Comment by hglattergotz** [comment]
As I mentioned earlier, I went to the "source" to get the answer on this issue. Based on the [answer](http://discuss.jsonapi.org/t/clarification-on-the-sparse-fieldset-response/221/11) by @ethanresnick on the forum I am now sure that this lib is not compliant in this particular case.

I don't need this feature for anything I am doing at the moment, but figured that neomerx/json-api is interested in being specification compliant so I pursued this even though the issue was closed.

**Comment by neomerx** [comment]
@hglattergotz You asked if specific behavior (exception of 'full-linkage' rule) is complaint with the spec. It doesn't mean other behavior (comply with 'full-linkage' rule) is not complaint.
As I understand the spec and the answer you mentioned it is allowed exception and implementation may choose either break or not 'full-linkage'.

I personally don't see why this behavior (called `bad practice` by your source BTW) should be preferred.

**Comment by neomerx** [comment]
If spec authors would say that implementations are not allowed to follow 'full linkage' rule it would be an argument.

**Comment by ethanresnick** [comment]
> It doesn't mean other behavior (comply with 'full-linkage' rule) is not complaint.
> As I understand the spec and the answer you mentioned it is allowed exception and implementation may choose either break or not 'full-linkage'.

I haven't followed this thread closely (it's long and I was just tagged) but, to hopefully clear things up: if you're including related resources, and sparse fieldsets is in use, and the relationship for the included resources isn't specified list of fields to return, then you MUST omit the relationship field that contains the linkage--this is just about conforming to sparse fieldsets--which means you then will be breaking the full linkage requirement.

**Comment by neomerx** [comment]
@hglattergotz  @lindyhopchris @ethanresnick  I've summed up opened questions in #105 

Please feel free to comment and send PRs with tests and code fixes
